src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,118 +291,188 @@ class HandshakeALotClosure : public HandshakeClosure {
+  // If we are at a safepoint we will evaluate all the operations that
+  // follow that also require a safepoint
+  log_debug(vmthread)("Evaluating %s %s VM operation: %s",
+                       prev_vm_operation ? "nested" : "",
+                      _cur_vm_operation->evaluate_at_safepoint() ? "safepoint" : "non-safepoint",
+                      _cur_vm_operation->name());
+
+  bool end_safepoint = false;
+  if (_cur_vm_operation->evaluate_at_safepoint() &&
+      !SafepointSynchronize::is_at_safepoint()) {
+    SafepointSynchronize::begin();
+    if (_timeout_task != NULL) {
+      _timeout_task->arm();
+    }
+    end_safepoint = true;
+  }
 
-        SafepointSynchronize::begin();
+  evaluate_operation(_cur_vm_operation);
 
-        if (_timeout_task != NULL) {
-          _timeout_task->arm();
-        }
+  if (end_safepoint) {
+    if (_timeout_task != NULL) {
+      _timeout_task->disarm();
+    }
+    SafepointSynchronize::end();
+  }
 
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  _cur_vm_operation = prev_vm_operation;
+}
 
-        if (_timeout_task != NULL) {
-          _timeout_task->disarm();
-        }
+void VMThread::wait_for_operation() {
+  MonitorLocker mu_queue(VMOperation_lock, Mutex::_no_safepoint_check_flag);
 
-        // Complete safepoint synchronization
-        SafepointSynchronize::end();
+  // Notify previous op done
+  mu_queue.notify();
 
-      } else {  // not a safepoint operation
-        log_debug(vmthread)("Evaluating non-safepoint VM operation: %s", _cur_vm_operation->name());
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  while (!should_terminate()) {
+    self_destruct_if_needed();
+    if (_next_vm_operation != NULL) {
+      return;
+    }
+    if (handshake_alot()) {
+      {
+        MutexUnlocker mul(VMOperation_lock);
+        HandshakeALotClosure hal_cl;
+        Handshake::execute(&hal_cl);
+      }
+      // When we unlocked above someone might have setup a new op.
+      if (_next_vm_operation != NULL) {
+        return;
       }
     }
+    assert(_next_vm_operation == NULL, "Must be");
+    assert(_cur_vm_operation  == NULL, "Must be");
 
-    //
-    //  Notify (potential) waiting Java thread(s)
-    { MonitorLocker mu(VMOperationRequest_lock, Mutex::_no_safepoint_check_flag);
-      mu.notify_all();
+    cleanup_safepoint_alot();
+    if (_next_vm_operation != NULL) {
+      return;
+    }
+
+    mu_queue.notify();
+    mu_queue.wait(GuaranteedSafepointInterval);
+  }
+}
+
+void VMThread::loop() {
+  assert(_cur_vm_operation == NULL, "no current one should be executing");
+
+  SafepointSynchronize::init(_vm_thread);
+
+  // Need to set a calling thread for ops not passed
+  // via the normal way.
+  cleanup_op.set_calling_thread(_vm_thread);
+  safepointALot_op.set_calling_thread(_vm_thread);
+
+  while (true) {
+    wait_for_operation();
+    if (!should_terminate()) {
```

Comment:
This is a trailing block. So it might just be `do { ... } while (!should_terminate())`?

Reply:
Fixing